### PR TITLE
Add of test determining if the function/method name is a dummy one in…

### LINKED ISSUE #1369: Allow dummy function to be redefined
Some services allow us to use decorator to register callbacks, to avoid creating different function and passing them into the service. Like `functools.singledispatch`:
```python
from functools import singledispatch

@singledispatch
def func(arg):
    pass

@func.register(int)
def _(arg):
    pass

@func.register(str)
def _(arg):
    pass
```
This has been addressed in #1032.

### Steps to reproduce
However, it doesn't work if we are creating our own registry:
```python
callbacks = {}

def register(name):
	def inner(func):
		callbacks[name] = func
	return inner

@register("a")
def dummy(arg):
    pass

@register("b")
def dummy(arg):
    pass
```

### Current behavior
```
E: 13, 0: function already defined line 9 (function-redefined)
```

### Expected behavior
No function-redefined warning. Probably just allow functions matching "dummy-variables-rgx" to be redefined.

### pylint --version output
```
No config file found, using default configuration
pylint 1.7.0,
astroid 1.5.0
Python 3.6.0 (v3.6.0:41df79263a11, Dec 23 2016, 07:18:10) [MSC v.1900 32 bit (In
tel)]
```

P.S. It doesn't work without decorator either, but I guess this is not a valid use case.
```python
from functools import singledispatch

@singledispatch
def func(arg):
    pass

def _(arg):
    pass
func.register(int)(_)

def _(arg):
    pass
func.register(str)(_)
```

### PR DESCRIPTION
**Pull request by hippo91**
…side _check_redefinition method. If it is, then do not emit function-redefined message. Add of corresponding test, ChangeLog and whatsnew entries.
Fixes #1369

### CODE REVIEW

File: pylint/checkers/base.py [hunk 1]
```diff
@@ -605,8 +606,13 @@ def _check_redefinition(self, redeftype, node):
         """check for redefinition of a function / method / class name"""
         defined_self = node.parent.frame()[node.name]
         if defined_self is not node and not astroid.are_exclusive(node, defined_self):
-            self.add_message('function-redefined', node=node,
-                             args=(redeftype, defined_self.fromlineno))
+            # before emitting a message check that function/method name doesn't not
+            # correspond to dummy names.
+            dummy_variables_rgx = lint_utils.get_global_option(
+                self, 'dummy-variables-rgx', default=None)
+            if dummy_variables_rgx.match(defined_self.name) is None:
```
Comment (rogalski) [inline pylint/checkers/base.py]: what if `dummy_variables_rgx` is None?
Also, we can simplify `... is None` check to `if not ...`

File: ChangeLog [hunk 2]
```diff
@@ -102,6 +102,10 @@ What's New in Pylint 1.8?
 
       Close #574
 
+    * Disabling ``function-redefined```message for function or method with
```
Comment (rogalski) [inline ChangeLog]: Mark-up is mixed up (two opening \`, three closing \`)
Maybe we can come up with more expressive description of change? It shouldn't mention disabling, as it'll introduce confusion with `pylint: disable=...` comments.

Maybe something along the lines of:

``function-redefined`` is no longer emitted for functions and methods which name matches dummy variable name regular expression.

Comment (hippo91) [reply ChangeLog]: Ok. Thanks @rogalski .

File: ChangeLog [hunk 3]
```diff
@@ -102,10 +102,15 @@ What's New in Pylint 1.8?
 
       Close #574
 
+    * ``function-redefined`` message is no longer emitted for functions and
+      methods wich names matches dummy variable name regular expression.
```
Comment (rogalski) [inline ChangeLog]: wich -> which

File: pylint/checkers/base.py [hunk 4]
```diff
@@ -605,6 +606,12 @@ def _check_redefinition(self, redeftype, node):
         """check for redefinition of a function / method / class name"""
         defined_self = node.parent.frame()[node.name]
         if defined_self is not node and not astroid.are_exclusive(node, defined_self):
+            # before emitting a message check that function/method name doesn't not
```
Comment (rogalski) [inline pylint/checkers/base.py]: commit is redundant - code says exact same thing

File: doc/whatsnew/1.8.rst [hunk 5]
```diff
@@ -276,5 +276,8 @@ Other Changes
     if 'bar' in (foo or {}):
       pass
 
+* Redefinition of dummy function is now possible. No ``function-redefined`` message will be emitted anymore if
```
Comment (rogalski) [inline doc/whatsnew/1.8.rst]: This one is very nitpicky (and I'm not a native speaker so I may be wrong), but:

_No message will be emitted_ seems slightly off comparing to _Message won't be emitted_.

Same goes to _if dummy functions are redefined_ - _**when** dummy functions are redefined_ seems slightly better.

**Pull request merged**